vllm/compilation/fusion_attn.py

```diff
@@ -166,4 +325,81 @@ def __call__(self, graph: torch.fx.graph.Graph) -> None:
+        make the QuantAttentionQuantPattern fusion pass works.
+        1. Caching is used to store the fusion status when fx graph is cached
+        2. Finding the host output_scale offline for Flashinfer TRTLLM attn API
+        """
+
+        # Init cache. This cache will save the layer fusion info,
+        # e.g. fused -> 1, not fused -> 0.
+        # This info is required at runtime to determined the attn kernel
+        # selection. Without this attn_fusion cache, the fusion status is
+        # unknown when the fx graph is cached, since the cached compiled
+        # graph will be loaded and the fusion pass will be skipped.
+        disabled_cache = envs.VLLM_DISABLE_COMPILE_CACHE
+        if not disabled_cache:
+            cache_dir = os.getenv("TORCHINDUCTOR_CACHE_DIR")
+            assert cache_dir is not None
+            assert os.path.exists(cache_dir)
+            cache_dir = os.path.join(cache_dir, "attn_fusion")
+            os.makedirs(cache_dir, exist_ok=True)
+
+        # Trying to find the input_scale from o_proj layer, which will be
+        # used as attn layer output_scale for the fusion. Currently using
+        # the layer name prefix matching to find the corresponding layers
+        # within the same decoder.
+        #
+        # This is mainly because Flashinfer TRTLLM attn currently only
+        # supports host scalar scale instead of device tensor scale, and
+        # a device tensor is not allowed to transfer to host at runtime
+        # with full cuda graph enabled.
+        input_scale = {}  # map from layer prefix name to input_scale
+        o_proj_pattern = re.compile(r"(.*layers\.\d+\.self_attn)\.o_proj$")
+        attn_pattern = re.compile(r"(.*layers\.\d+\.self_attn)\.attn$")
+
+        # Collect the input_scale for each self_attn.o_proj layer
+        for layer_name, layer in get_layers_from_vllm_config(
+                config, LinearBase).items():
+            match = o_proj_pattern.search(layer_name)
+            if match and hasattr(layer, "input_scale"):
+                input_scale[match.group(1)] = layer.input_scale.item()
+
+        # Only register the layer that the input_scale could be found
+        for layer_name, layer in get_layers_from_vllm_config(
+                config, Attention).items():
+            match = attn_pattern.search(layer_name)
+            # only register the pass when the input_scale is found
+            if not match or not input_scale.get(match.group(1)):
+                logger.debug(
+                    "Cannot find o_proj layer or input_scale for fusing %s",
+                    layer_name)
+                continue
+            layer._o_scale_float = input_scale[match.group(1)]
+
+            # load the fusion status into layer fused_quant
+            cache_file: Optional[str] = None
+            if not disabled_cache:
+                assert cache_dir is not None
+                cache_file = os.path.join(cache_dir, layer_name)
```

Comment:
I am worried about using layer name as the file name, because the layer name may contain invalid characters for file paths.

Is it possible to put all the layers' result in the same file (say, `os.path.join(cache_dir, "QuantAttentionQuantPattern.json")`, in the format of something like:

```
{"layer_name_1": True, "layer_name_2": False, "layer_name_3": True}
```

so that we only open this file once per run instead of opening it again and again for each layer?

Reply:
In `extra_check()`, we can read the file, load the json into dict, add the new layer into the dict, and then write back the dict as json?

Reply:
discussed offline. The author prefers the current approach.

My only requirement is to escape the invalid characters using `re.sub(r'[^\w_.-]', '_', layer_name)`